### DISCUSSION TITLE
Should enum types be supported?

### DISCUSSION THREAD

**Discussion by jeffreyvr** [spatie/laravel-livewire-wizard#118] [Q&A]
If you use enum types on StepComponent class properties, you'll get an error when the step state is being set.

I run into this when I go to the next step, and then back (using `nextStep` and `previousStep`). Example error:

```
Cannot assign string to property App\Livewire\Wizards\Steps\SubscriptionSelection::$paymentInterval of type App\Enums\ SubscriptionPaymentInterval
```

Part of my component:
```php
class SubscriptionSelection extends StepComponent
{
    public SubscriptionPaymentInterval $paymentInterval = SubscriptionPaymentInterval::YEARLY;
}
```

Now this could be fixed by if we hook into the `setStepState` method, and check if any of the properties are enums. This concept code fixes the issue (part of `WizardComponent`):

```php
public function setStatePropertyTypes($step, array $state = [])
{
    $stepClass = app(ComponentRegistry::class)->getClass($step);

    $reflectionClass = (new \ReflectionClass($stepClass));

    foreach($reflectionClass->getProperties() as $property) {
        $propertyName = $property->getName();
        $propertyType = $property->getType();

        if($propertyType === null || ! isset($state[$propertyName])) {
            continue;
        }

        if($propertyType->getName() !== gettype($state[$propertyName])) {
            if(! class_exists($propertyType->getName())) {
                continue;
            }

            if((new \ReflectionClass($propertyType->getName()))->isEnum()) {
                $state[$propertyName] = $propertyType->getName()::from($state[$propertyName]);
            }
        }
    }

    return $state;
}

public function setStepState(string $step, array $state = []): void
{
    if (! $this->stepNames()->contains($step)) {
        throw StepDoesNotExist::doesNotHaveState($step);
    }

    $this->allStepState[$step] = $this->setStatePropertyTypes($step, $state);
}
```

Is this something that should work, or this a potential rabbit hole this package shouldn't get into? I could always remove the types and cast them later on.

I'd be willing to make a PR, but would understand it too if it is not wanted.

**Comment by ju5t** [comment]
> or this a potential rabbit hole this package shouldn't get into?

Probably :)

Because it's not just enums. I think the best approach here would be to hook into Livewire's hydration/synths somehow.

**Reply by jeffreyvr** [reply]
>  I think the best approach here would be to hook into Livewire's hydration/synths somehow.

That would be a nice way. I did take a quick look at it, but might be a challenge I think.
